### ISSUE TITLE
Cron Update fails (again)

### ISSUE THREAD

**Issue by wmtech-1** [netdata/netdata#5208]
##### Bug report summary

Cron Log Output:

2019-01-18 03:47:05 (32.0 MB/s) - ‘netdata-latest.tar.gz’ saved [3497026/3497026]

./netdata-latest.tar.gz: OK
Fri Jan 18 03:47:05 UTC 2019 : INFO:  Re-installing netdata...
/etc/cron.daily/netdata-updater: line 94: ./netdata-installer.sh: Permission denied
Fri Jan 18 03:47:05 UTC 2019 : ERROR:  FAILED TO UPDATE NETDATA : FAILED TO COMPILE/INSTALL NETDATA

##### OS / Environment

Linux 3.10.0-957.1.3.el7.x86_64

##### Netdata version

Most recent before Update

##### Component Name

Update

##### Steps To Reproduce

Run cron update

##### Expected behavior

Finish update successfully.

Please take a look at this problem and be more careful with changes regarding updates. This is the third issue with updates within a few months.

**Comment by paulfantom** [comment]
We are very sorry for inconvenience. After last time we wrote some tests which should catch such problems before they land on master branch, but unfortunately somehow this bug went through. This time we changed where update packages are taken from to prevent such issues in the future. This breaks previous model of distributing updates based on git repository and from now on it will fetch a nightly package from Google Cloud Storage when needed.

I need some more info about your environment to debug it. It looks like you are using CentOS7 and updater cannot run installer script due to some permission issue. Are you by any chance using SELinux in enforcing mode with ruleset blocking execution access to files in `/tmp`?

**Comment by paulfantom** [comment]
I think this might be fixed in https://github.com/netdata/netdata/pull/5209, which means that next autoupdate should work ok, without any action needed from your side.

**Comment by paulfantom** [comment]
However if this is still unsolved, could you attach latest `/tmp/netdata-updater.log.XXXXXX` (Xs are random characters), those files store outputs of every command which was run by updater.

Also it would be nice to get one from when update failed. If you don't want posting them here, you can send it to me via email pawel@netdata.cloud.

**Comment by Daniel15** [comment]
I'm seeing this on my Debian systems:
```
/etc/cron.daily/netdata-updater:
run-parts: failed to exec /etc/cron.daily/netdata-updater: Exec format error
run-parts: /etc/cron.daily/netdata-updater exited with return code 1
```

`/etc/cron.daily/netdata-updater` is a symlink to `/usr/src/netdata.git/netdata-updater.sh`, which appears to be an empty file?

**Comment by Daniel15** [comment]
How do I fix it? What should the contents of the file be?

**Comment by paulfantom** [comment]
By default (when netdata was insalled without any prefixed directories) it would look like a file from https://github.com/netdata/netdata/blob/master/packaging/installer/netdata-updater.sh with `THIS_SHOULD_BE_REPLACED_BY_INSTALLER_SCRIPT` replaced to `/etc/netdata/.environment`.

**Comment by paulfantom** [comment]
Essentially you can run this as root:
```bash
UPDATER="/etc/cron.daily/netdata-updater"
rm "$UPDATER"
curl https://raw.githubusercontent.com/netdata/netdata/master/packaging/installer/netdata-updater.sh | sed 's@THIS_SHOULD_BE_REPLACED_BY_INSTALLER_SCRIPT@/etc/netdata/.environment@' > $UPDATER
chown root:root $UPDATER
chmod 0755 $UPDATER
```

**Comment by wmtech-1** [comment]
I did find our problem. All our /tmp directories are mounted with the noexec option. So it is not possible to run a script there...

Since it is not uncommon to protect /tmp directories I strongly recommend to create a directory within the netdata directory (e.g. /usr/share/netdata/tmp) and use that directory for the installation files (eg: "mktemp -d --tmpdir=/usr/share/netdata/tmp") in the updater script.

Such a change will make your installer much less prone to errors.

**Comment by paulfantom** [comment]
I have added a detection mechanism in #5232. After that updater will be downloading package into `/opt` when `/tmp` is not available. In the future we might change this location to something more suitable, like `/usr/local/share/netdata/tmp`, but right now we need something that works on every platform.

**Comment by ktsaou** [comment]
This happened to a netdata demo site today.

Received cron report:

```sh
/etc/cron.daily/netdata-updater:
run-parts: failed to exec /etc/cron.daily/netdata-updater: Exec format error
run-parts: /etc/cron.daily/netdata-updater exited with return code 1
```

These are the files:

```sh
# ls -l /etc/cron.daily/netdata-updater 
lrwxrwxrwx 1 root root 39 Jan  2 06:25 /etc/cron.daily/netdata-updater -> /usr/src/netdata.git/netdata-updater.sh

# ls -l /usr/src/netdata.git/netdata-updater.sh
-rwxr-xr-x 1 root root 0 Jan 18 06:25 /usr/src/netdata.git/netdata-updater.sh
```

So, `netdata-updater.sh` is an empty file.

This is the version of netdata that did this:

```sh
# cd /usr/src/netdata.git/
cdn77 netdata.git # git describe
v1.12.0-rc2-52-g61e3100
```

**Comment by wmtech-1** [comment]
Update is working without problems now. Thank you for the quick fix!

Would be a good idea to include this fix to the installer also. Currently netdata does not install at machines where the /tmp filesystem does not allow execute files.

**Comment by paulfantom** [comment]
I don't see why we would need to do this in installer. I agree that this should be included in all wrappers around installer (like updater or kickstart), but not in installer per se. Installer is meant to be run from a directory it is in, so if repo was put into `/tmp` mounted with `noexec`, then there is nothing we can do about it as installer cannot be run.

Fix for this is to use updater with detection of `noexec` for `/tmp` (should be done by https://github.com/netdata/netdata/pull/5232) and the same mechanism applied to `kickstart.sh`.

**Comment by cakrit** [comment]
For people who can't get the fix in #5232 because they have the `netdata-updater` that only uses `/tmp`, a simple workaround is in #5289.

I believe we can safely close this one now, please a new issue if you notice anything else.
